# NumPy: Handling Numerical Data

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Danselem/brics_astro/blob/main/Week2/01_numpy.ipynb)

Welcome back! When working with astronomical data, you'll quickly find that you need to handle large sets of numbers – lists of star brightnesses, grids of pixel values in an image, arrays of simulation results, and much more.

While Python's built-in lists are versatile, they aren't the most efficient tool for performing mathematical operations on large collections of numbers. This is where **NumPy** (pronounced "Num-py") comes in.

NumPy is a fundamental library in Python that provides essential tools for working with arrays of numbers. It makes numerical calculations faster and easier to write compared to using standard Python lists and loops. Many other scientific Python libraries, like Pandas and Matplotlib, are built on top of NumPy.

In this notebook, we will introduce the core concepts of NumPy and demonstrate how you can use it to start handling your astronomical data more efficiently.

**Learning Objectives:**

*   Understand why NumPy is important for numerical tasks.
*   Learn about the core NumPy object: the `ndarray`.
*   Create NumPy arrays in different ways.
*   Access properties of NumPy arrays (shape, size, data type).
*   Select and modify parts of arrays using indexing and slicing.
*   Perform basic mathematical operations on entire arrays.
*   Use common NumPy functions for calculations.
*   See examples of how NumPy is used in astronomy.

**Prerequisites:**

*   Basic familiarity with Python syntax (variables, data types, lists, basic loops).
*   Familiarity with Jupyter Notebooks.

**Note**
If you are running this jupyter notebook from Colab, then run the next cell by pressing `SHIFT+ENTER` to install the required packages for this notebook. Otherwise, skip the next cell.

In [ ]:
!pip install numpy pandas matplotlib

## Why Numpy
NumPy is Essential for Numerical Tasks.

Imagine you have a list of star brightness measurements defined as `star_brightnesses`.

In [1]:
star_brightnesses = [10.5, 12.1, 9.8, 11.5, 13.0]

If you wanted to multiply each brightness by 2 (maybe converting units), with lists you might do:

In [2]:
doubled_brightnesses_list = []
for brightness in star_brightnesses:
    doubled_brightnesses_list.append(brightness * 2)

print("Using Python list and loop:", doubled_brightnesses_list)

Using Python list and loop: [21.0, 24.2, 19.6, 23.0, 26.0]


This works, but for very large datasets (like millions of pixels in an image), this loop can be slow.

NumPy allows us to do this much more simply and efficiently:

In [3]:
import numpy as np # Standard way to import NumPy

# Convert the list to a NumPy array
star_brightnesses_np = np.array(star_brightnesses)

# Multiply the *entire array* by 2
doubled_brightnesses_np = star_brightnesses_np * 2

print("Using NumPy:", doubled_brightnesses_np)

Using NumPy: [21.  24.2 19.6 23.  26. ]


This ability to perform operations on whole arrays at once is called `vectorisation`, and it is one reason NumPy is efficient for numerical calculations.

## The `ndarray`: NumPy's Core Object

The central feature of NumPy is the `ndarray` object. An `ndarray` (N-dimensional array) is a grid of values, all of the same data type. You can think of it like a table (2 dimensions), or even a cube of numbers (3 dimensions), or more!

*   **Homogeneous:** All elements in a NumPy array must be of the same data type (e.g., all integers, all floating-point numbers). This is a key difference from Python lists and helps with efficiency.
*   **N-dimensional:** Can have any number of dimensions (1D for a list of values, 2D for a table, 3D for image cubes, etc.).


### Example 1: Creating Arrays

In [4]:
# 1. From a Python list or tuple:
# A 1-dimensional array (like a list)
star_magnitudes = [2.5, 1.8, 0.1, -1.4]
magnitude_array = np.array(star_magnitudes)
print("1D Array from list:", magnitude_array)

# A 2-dimensional array (like a table)
# Representing [RA, Dec, Magnitude] for two stars
star_data = [
    [101.28, -16.71, -1.46], # Sirius
    [37.95, 89.26, -2.08]   # Polaris
]
star_array_2d = np.array(star_data)
print("\n2D Array from list of lists:")
print(star_array_2d)

1D Array from list: [ 2.5  1.8  0.1 -1.4]

2D Array from list of lists:
[[101.28 -16.71  -1.46]
 [ 37.95  89.26  -2.08]]


In [5]:
# 2. Using built-in NumPy functions:
# An array of zeros (useful for initializing data)
zeros_array = np.zeros(5) # A 1D array of 5 zeros
print("\nArray of zeros:", zeros_array)

# A 2x3 array of ones
ones_array = np.ones((2, 3)) # Shape is given as a tuple (rows, columns)
print("\nArray of ones (2x3):")
print(ones_array)

# An array with values from 0 up to (but not including) 10
arange_array = np.arange(10)
print("\nArray using arange(10):", arange_array)

# An array of 5 evenly spaced values between 0 and 1 (inclusive)
linspace_array = np.linspace(0, 1, 5)
print("\nArray using linspace(0, 1, 5):", linspace_array)


Array of zeros: [0. 0. 0. 0. 0.]

Array of ones (2x3):
[[1. 1. 1.]
 [1. 1. 1.]]

Array using arange(10): [0 1 2 3 4 5 6 7 8 9]

Array using linspace(0, 1, 5): [0.   0.25 0.5  0.75 1.  ]


## Array Attributes: What an Array Knows About Itself

NumPy arrays have several useful attributes that tell you about their structure and contents.

*   `.shape`: A tuple indicating the size of each dimension (rows, columns, etc.).
*   `.size`: The total number of elements in the array.
*   `.ndim`: The number of dimensions (e.g., 1 for a 1D array, 2 for a 2D array).
*   `.dtype`: The data type of the elements in the array (e.g., `int64`, `float64`).


### Example 2: Array Attributes
Let us use the arrays we created earlier:

In [6]:
magnitude_array = np.array([2.5, 1.8, 0.1, -1.4])
star_array_2d = np.array([[101.28, -16.71, -1.46], [37.95, 89.26, -2.08]])

print("--- Attributes of 1D Array ---")
print("Array:", magnitude_array)
print("Shape:", magnitude_array.shape) # (4,) means 1 dimension with 4 elements
print("Size:", magnitude_array.size)   # Total elements
print("Number of dimensions:", magnitude_array.ndim)
print("Data type:", magnitude_array.dtype)

print("\n--- Attributes of 2D Array ---")
print("Array:")
print(star_array_2d)
print("Shape:", star_array_2d.shape) # (2, 3) means 2 rows, 3 columns
print("Size:", star_array_2d.size)
print("Number of dimensions:", star_array_2d.ndim)
print("Data type:", star_array_2d.dtype) # Notice NumPy often chooses float by default for mixed numbers

--- Attributes of 1D Array ---
Array: [ 2.5  1.8  0.1 -1.4]
Shape: (4,)
Size: 4
Number of dimensions: 1
Data type: float64

--- Attributes of 2D Array ---
Array:
[[101.28 -16.71  -1.46]
 [ 37.95  89.26  -2.08]]
Shape: (2, 3)
Size: 6
Number of dimensions: 2
Data type: float64


## Array Indexing and Slicing: Accessing Elements

Just like with Python lists, you can access individual elements or portions of a NumPy array using indexing and slicing.

*   **Indexing:** Accessing a single element using its position (index). Remember Python is 0-indexed!
*   **Slicing:** Accessing a range of elements.

For 2D arrays, you use a comma to separate the row index/slice from the column index/slice: `array[row_index_or_slice, column_index_or_slice]`.

### Example 3: Indexing and Slicing

In [7]:
star_data = np.array([
    [101.28, -16.71, -1.46],  # Row 0 (Sirius)
    [37.95, 89.26, -2.08],    # Row 1 (Polaris)
    [217.42, -62.68, 0.13]    # Row 2 (Proxima Centauri)
])
print("Original Array:")
print(star_data)

print("\n--- Indexing ---")
# Get the value at row 1, column 2 (Polaris's magnitude)
polaris_magnitude = star_data[1, 2]
print("Polaris's magnitude:", polaris_magnitude)

Original Array:
[[ 1.0128e+02 -1.6710e+01 -1.4600e+00]
 [ 3.7950e+01  8.9260e+01 -2.0800e+00]
 [ 2.1742e+02 -6.2680e+01  1.3000e-01]]

--- Indexing ---
Polaris's magnitude: -2.08


In [8]:
# Get the value at row 0, column 0 (Sirius's RA)
sirius_ra = star_data[0, 0]
print("Sirius's RA:", sirius_ra)

# Get an entire row (Row 2 - Proxima Centauri data)
proxima_row = star_data[2, :] # ':' means select all columns in this row
print("\nProxima Centauri's data (row 2):", proxima_row)

# Get an entire column (Column 1 - Declination for all stars)
declination_column = star_data[:, 1] # ':' means select all rows in this column
print("Declinations of all stars (column 1):", declination_column)


print("\n--- Slicing ---")
# Get the first two rows
first_two_rows = star_data[0:2, :] # From row 0 up to (not including) row 2, all columns
print("First two rows:")
print(first_two_rows)

Sirius's RA: 101.28

Proxima Centauri's data (row 2): [ 2.1742e+02 -6.2680e+01  1.3000e-01]
Declinations of all stars (column 1): [-16.71  89.26 -62.68]

--- Slicing ---
First two rows:
[[101.28 -16.71  -1.46]
 [ 37.95  89.26  -2.08]]


In [9]:
# Get columns 1 and 2 (Dec and Magnitude) for all stars
dec_mag_columns = star_data[:, 1:3] # All rows, from column 1 up to (not including) column 3
print("Declination and Magnitude columns:")
print(dec_mag_columns)

# Get the RA and Dec for the first two stars
first_two_stars_ra_dec = star_data[0:2, 0:2]
print("RA and Dec for first two stars:")
print(first_two_stars_ra_dec)

Declination and Magnitude columns:
[[-16.71  -1.46]
 [ 89.26  -2.08]
 [-62.68   0.13]]
RA and Dec for first two stars:
[[101.28 -16.71]
 [ 37.95  89.26]]


In [10]:
# Modifying an element using indexing
star_data[1, 2] = -1.80 # Update Polaris's magnitude
print("\nArray after updating Polaris's magnitude:")
print(star_data)


Array after updating Polaris's magnitude:
[[ 1.0128e+02 -1.6710e+01 -1.4600e+00]
 [ 3.7950e+01  8.9260e+01 -1.8000e+00]
 [ 2.1742e+02 -6.2680e+01  1.3000e-01]]


## Basic Operations on Arrays

You can perform mathematical operations directly on entire arrays. NumPy applies the operation element-wise (to each element) or uses broadcasting when operating between arrays of different shapes or between an array and a single number.

### Example 4: Array Arithmetic

In [11]:
magnitudes = np.array([2.5, 1.8, 0.1, -1.4])
offsets = np.array([0.1, 0.1, 0.1, 0.1]) # An array of same shape

# Addition: Add an offset to all magnitudes
shifted_magnitudes = magnitudes + offsets
print("Magnitudes + offset array:", shifted_magnitudes)

# Subtraction:
subtracted_values = magnitudes - offsets
print("Magnitudes - offset array:", subtracted_values)

Magnitudes + offset array: [ 2.6  1.9  0.2 -1.3]
Magnitudes - offset array: [ 2.4  1.7  0.  -1.5]


In [12]:
# Multiplication: Double the magnitudes
doubled_magnitudes = magnitudes * 2 # Multiplying by a single number (broadcasting)
print("Magnitudes * 2:", doubled_magnitudes)

# Division: Divide magnitudes by 4
divided_magnitudes = magnitudes / 4.0
print("Magnitudes / 4:", divided_magnitudes)

# Exponentiation: Calculate 10 raised to the power of -0.4 * magnitude
# This is how you convert magnitudes back to relative flux/brightness
relative_fluxes = 10**(-0.4 * magnitudes)
print("Magnitudes converted to relative fluxes:", relative_fluxes)

Magnitudes * 2: [ 5.   3.6  0.2 -2.8]
Magnitudes / 4: [ 0.625  0.45   0.025 -0.35 ]
Magnitudes converted to relative fluxes: [0.1        0.19054607 0.91201084 3.63078055]


### Example 5: Array Comparisons

In [13]:
magnitudes = np.array([2.5, 1.8, 0.1, -1.4, 3.0, 0.5])

# Find stars brighter than magnitude 1.0 (remember smaller magnitude = brighter)
is_brighter_than_1 = magnitudes < 1.0
print("Is magnitude < 1.0?", is_brighter_than_1) # Returns a boolean array

Is magnitude < 1.0? [False False  True  True False  True]


In [14]:
# Use the boolean array for filtering (Fancy Indexing)
bright_stars = magnitudes[is_brighter_than_1]
print("Magnitudes of stars brighter than 1.0:", bright_stars)

Magnitudes of stars brighter than 1.0: [ 0.1 -1.4  0.5]


In [15]:
# Find stars with magnitude exactly equal to 0.1
is_mag_0_1 = magnitudes == 0.1
print("\nIs magnitude == 0.1?", is_mag_0_1)


Is magnitude == 0.1? [False False  True False False False]


In [16]:
# Find stars with magnitude greater than 0 AND less than 2
# Need parentheses for each comparison
is_between_0_and_2 = (magnitudes > 0) & (magnitudes < 2) # Use '&' for element-wise AND
print("\nIs magnitude between 0 and 2?", is_between_0_and_2)
print("Magnitudes between 0 and 2:", magnitudes[is_between_0_and_2])


Is magnitude between 0 and 2? [False  True  True False False  True]
Magnitudes between 0 and 2: [1.8 0.1 0.5]


### Example 6: Common NumPy Functions

In [17]:
stellar_masses = np.array([0.8, 1.2, 0.5, 2.0, 0.9]) # Masses in solar masses

# Sum: Calculate the total mass of these stars
total_mass = np.sum(stellar_masses)
print("Total mass:", total_mass)

# Mean: Calculate the average mass
average_mass = np.mean(stellar_masses)
print("Average mass:", average_mass)

Total mass: 5.4
Average mass: 1.08


In [18]:
# Max: Find the maximum mass
max_mass = np.max(stellar_masses)
print("Maximum mass:", max_mass)

# Min: Find the minimum mass
min_mass = np.min(stellar_masses)
print("Minimum mass:", min_mass)

Maximum mass: 2.0
Minimum mass: 0.5


In [19]:
# Other useful functions (element-wise):
orbital_phases_radians = np.array([0, np.pi/2, np.pi, 3*np.pi/2, 2*np.pi]) # Angles in radians
print("\nOrbital phases (radians):", orbital_phases_radians)
print("Cosine of phases:", np.cos(orbital_phases_radians)) # Cosine applied to each element
print("Sine of phases:", np.sin(orbital_phases_radians))   # Sine applied to each element


Orbital phases (radians): [0.         1.57079633 3.14159265 4.71238898 6.28318531]
Cosine of phases: [ 1.0000000e+00  6.1232340e-17 -1.0000000e+00 -1.8369702e-16
  1.0000000e+00]
Sine of phases: [ 0.0000000e+00  1.0000000e+00  1.2246468e-16 -1.0000000e+00
 -2.4492936e-16]


## Putting it Together: Astronomy Examples with NumPy

Let's look at some more complete examples applying NumPy to astronomical problems.

### Example 7: Analyzing a Simple Star Catalog Array
Imagine this array represents [Temperature (K), Luminosity (Solar Lum), Distance (pc)] for several stars

In [22]:
star_catalog_data = np.array([
    [5778, 1.0, 4.85],    # Sun
    [9940, 83.0, 2.64],   # Sirius A
    [4500, 0.3, 10.0],    # Alpha Centauri B
    [25000, 10000.0, 100.0] # O-type star example
])
print(star_catalog_data)

[[5.778e+03 1.000e+00 4.850e+00]
 [9.940e+03 8.300e+01 2.640e+00]
 [4.500e+03 3.000e-01 1.000e+01]
 [2.500e+04 1.000e+04 1.000e+02]]


In [23]:
# Select the Temperature column
temperatures = star_catalog_data[:, 0]
print("Temperatures:", temperatures)

# Select the Luminosity column
luminosities = star_catalog_data[:, 1]
print("Luminosities:", luminosities)

Temperatures: [ 5778.  9940.  4500. 25000.]
Luminosities: [1.0e+00 8.3e+01 3.0e-01 1.0e+04]


In [24]:
# Select the Distance column
distances_pc = star_catalog_data[:, 2]
print("Distances (pc):", distances_pc)

# Calculate the average temperature of these stars
average_temp = np.mean(temperatures)
print(f"\nAverage temperature: {average_temp:.1f} K")

Distances (pc): [  4.85   2.64  10.   100.  ]

Average temperature: 11304.5 K


In [25]:
# Calculate the total luminosity
total_luminosity = np.sum(luminosities)
print(f"Total luminosity: {total_luminosity:.1f} Solar Lum")

# Convert distances from parsecs (pc) to light-years (ly) (1 pc ≈ 3.26 ly)
distances_ly = distances_pc * 3.26
print(f"Distances (ly): {distances_ly}")

Total luminosity: 10084.3 Solar Lum
Distances (ly): [ 15.811    8.6064  32.6    326.    ]


### Example 8: Simulating Brightness Decrease with Distance

Assume initial brightness at a reference distance is 100 (arbitrary units)

In [27]:
initial_brightness = 100.0

# Create an array of increasing distances
distances = np.array([1, 2, 3, 4, 5, 10]) # Distances in some unit

# Brightness decreases with the square of the distance (Brightness = Initial / Distance^2)
brightness_at_distance = initial_brightness / (distances ** 2)

print(f"Brightness at distances {distances}: \n {brightness_at_distance}")

Brightness at distances [ 1  2  3  4  5 10]: 
 [100.          25.          11.11111111   6.25         4.
   1.        ]


In [28]:
# Find distances where brightness is below a threshold (e.g., 5)
threshold = 5.0
is_below_threshold = brightness_at_distance < threshold
print(f"\nIs brightness below {threshold}? {is_below_threshold}")
print(f"Distances where brightness is below {threshold}: {distances[is_below_threshold]}")


Is brightness below 5.0? [False False False False  True  True]
Distances where brightness is below 5.0: [ 5 10]


### Example 9: Simple 2D Image Data (Grayscale)

Imagine a small 3x3 pixel image represented as a 2D array of pixel intensity values


In [29]:
image_data = np.array([
    [10, 20, 15],
    [30, 25, 35],
    [5, 15, 10]
])
print("Image Data (3x3 array):")
print(image_data)
print("Shape of image:", image_data.shape)

Image Data (3x3 array):
[[10 20 15]
 [30 25 35]
 [ 5 15 10]]
Shape of image: (3, 3)


In [30]:
# Accessing a specific pixel (e.g., the center pixel)
center_pixel = image_data[1, 1]
print("\nCenter pixel value:", center_pixel)


Center pixel value: 25


In [31]:
# Get a row of pixels (e.g., the middle row)
middle_row = image_data[1, :]
print("Middle row of pixels:", middle_row)

Middle row of pixels: [30 25 35]


In [32]:
# Get a column of pixels (e.g., the first column)
first_column = image_data[:, 0]
print("First column of pixels:", first_column)

First column of pixels: [10 30  5]


In [33]:
# Calculate the average pixel intensity in the image
average_intensity = np.mean(image_data)
print(f"\nAverage pixel intensity: {average_intensity:.2f}")


Average pixel intensity: 18.33


In [34]:
# Find pixels with intensity greater than 20
bright_pixels_mask = image_data > 20
print("Mask for bright pixels (>20):\n", bright_pixels_mask)

Mask for bright pixels (>20):
 [[False False False]
 [ True  True  True]
 [False False False]]


In [35]:
# Get the values of bright pixels
bright_pixel_values = image_data[bright_pixels_mask]
print("Values of bright pixels:", bright_pixel_values)

Values of bright pixels: [30 25 35]


## Saving and Loading NumPy Arrays

In astronomy, you'll often need to save arrays you've created or manipulated to a file so you can use them later or share them with others. You'll also need to load arrays that someone else has saved. NumPy provides convenient functions for this.

There are two main approaches for beginners:

1.  **NumPy's binary format (`.npy`):** This is the most efficient way to save and load NumPy arrays while preserving their exact data type and shape. The files are not human-readable.
2.  **Text format (`.txt`, `.csv`, etc.):** This saves the data in a human-readable format, often with values separated by commas (CSV) or spaces. These files can be opened by other programs (like spreadsheets) but might not preserve all NumPy-specific information.

Let's look at how to use `np.save()`, `np.load()`, `np.savetxt()`, and `np.loadtxt()`.

### Example 10: Saving and Loading with .npy format
Imagine this is an array of calculated star distances in light-years

In [41]:
star_distances_ly = np.array([4.37, 8.60, 10.0, 14.5, 36.0])

### Saving the array to a .npy file 
`np.save()` takes the filename (without extension, `.npy` is added automatically)
and the array you want to save.

In [42]:
np.save("my_star_distances", star_distances_ly) # This will create a file named "my_star_distances.npy"
print("Array saved to 'my_star_distances.npy'")

Array saved to 'my_star_distances.npy'


### Loading the array from the .npy file
`np.load()` takes the filename (including the `.npy` extension)

In [43]:
loaded_distances = np.load("my_star_distances.npy")
print("\nArray loaded from 'my_star_distances.npy':")
print(loaded_distances)


Array loaded from 'my_star_distances.npy':
[ 4.37  8.6  10.   14.5  36.  ]


In [38]:
# Verify it's a NumPy array and has the same data
print("Type of loaded data:", type(loaded_distances))
print("Loaded data matches original:", np.array_equal(star_distances_ly, loaded_distances))

Type of loaded data: <class 'numpy.ndarray'>
Loaded data matches original: True


### Example 11: Saving and Loading with Text Files (.txt or .csv)

Imagine this 2D array is data for a few exoplanets: [Mass (Earth), Radius (Earth), Orbital Period (days)]

In [44]:
exoplanet_table = np.array([
    [1.0, 1.0, 365.25],   # Earth
    [0.11, 0.53, 687.0],  # Mars
    [317.8, 11.2, 4331.0] # Jupiter
])
print("Exoplanet data to save:")
print(exoplanet_table)

Exoplanet data to save:
[[1.0000e+00 1.0000e+00 3.6525e+02]
 [1.1000e-01 5.3000e-01 6.8700e+02]
 [3.1780e+02 1.1200e+01 4.3310e+03]]


### Saving the array to a text file (e.g., CSV)
`np.savetxt()` is used for text files.

It takes the filename, the array, and optional arguments like 'delimiter'

In [45]:
np.savetxt("exoplanet_data.csv", exoplanet_table, delimiter=",") # Use comma as separator
print("\nArray saved to 'exoplanet_data.csv' (CSV format)")


Array saved to 'exoplanet_data.csv' (CSV format)


### Loading the array from the text file
`np.loadtxt()` is used for loading text files.

It takes the filename and optional arguments like 'delimiter'.

In [46]:
loaded_exoplanet_data = np.loadtxt("exoplanet_data.csv", delimiter=",")
print("\nArray loaded from 'exoplanet_data.csv':")
print(loaded_exoplanet_data)

# Verify it's a NumPy array and has the same data
print("Type of loaded data:", type(loaded_exoplanet_data))
print("Loaded data matches original:\n", np.array_equal(exoplanet_table, loaded_exoplanet_data))

# You can open 'exoplanet_data.csv' in a text editor or spreadsheet program to see the data.


Array loaded from 'exoplanet_data.csv':
[[1.0000e+00 1.0000e+00 3.6525e+02]
 [1.1000e-01 5.3000e-01 6.8700e+02]
 [3.1780e+02 1.1200e+01 4.3310e+03]]
Type of loaded data: <class 'numpy.ndarray'>
Loaded data matches original:
 True


### Example 12: Loading Text File with Headers/Comments

Sometimes, astronomical data files have header lines or comments.
`np.genfromtxt()` can skip these.

Imagine a file 'star_observations.txt' like this:

```code
# Observation Data - 2023-10-27
# Star Name, Flux, Error
  Alpha, 123.45, 1.2
  Beta, 56.78, 0.8
  Gamma, 234.56, 2.1
```


In [47]:
# Create a dummy file for this example:
file_content = """# Observation Data - 2023-10-27
# Star Name, Flux, Error
Alpha, 123.45, 1.2
Beta, 56.78, 0.8
Gamma, 234.56, 2.1
"""

In [48]:
with open("star_observations.txt", "w") as f:
    f.write(file_content)
print("Created dummy file 'star_observations.txt'")

Created dummy file 'star_observations.txt'


In [50]:
# --- Loading data, skipping the first 2 lines (header/comments) ---
# use skiprows parameter
loaded_obs_data = np.genfromtxt(
    "star_observations.txt", 
    delimiter=",", 
    skip_header=2,       # same effect as skiprows
    dtype=None,          # auto-detect data types
    encoding='utf-8'     # handle string decoding
)
print("\nData loaded from 'star_observations.txt' (skipping 2 rows):")
print(loaded_obs_data)

# Note: loadtxt is best for purely numerical data. For mixed data (like star names and numbers),
# pandas read_csv (which uses numpy internally) is often more convenient.


Data loaded from 'star_observations.txt' (skipping 2 rows):
[('Alpha', 123.45, 1.2) ('Beta',  56.78, 0.8) ('Gamma', 234.56, 2.1)]


## Exercises

Practice makes perfect! Try these exercises to solidify your understanding of NumPy basics:

1. Create a 1D NumPy array named `planet_radii` containing the approximate radii of the inner planets in Earth radii: `[0.38, 0.95, 1.0, 0.53]`.
2. For the `planet_radii` array you created, print its `.shape`, `.size`, and `.ndim`.
3. From the `planet_radii` array, access and print only the radius of Earth (which is the third value in the sequence).
4. From the `planet_radii` array, access and print only the radius of Mars (the last value).
5. Create a new NumPy array named `planet_diameters` by multiplying the `planet_radii` array by 2. Print the `planet_diameters` array.
6. Using the `planet_radii` array, create a boolean array that is `True` for planets with a radius less than 0.6 Earth radii. Use this boolean array to print the radii of only the smaller planets.
7. Calculate and print the average radius from the `planet_radii` array using a NumPy function.
8. Create a 2D NumPy array named `star_distances` with shape (3, 2). Fill it with hypothetical data where each row is a star, the first column is distance in parsecs (pc), and the second column is distance in light-years (ly). You can use values like `[[5.0, 16.3], [10.0, 32.6], [2.5, 8.15]]`.
9. From the `star_distances` array, print only the column containing distances in light-years.
10. From the `star_distances` array, find and print the maximum distance in the parsecs column.
11. Save the planet_radii array you created earlier `([0.38, 0.95, 1.0, 0.53])` to a .npy file named `inner_planet_radii.npy`. Then, load this file into a new variable and print it to verify.
12. Save the star_distances 2D array you created earlier `([[5.0, 16.3], [10.0, 32.6], [2.5, 8.15]])` to a CSV file named `star_distances.csv` using a comma as the delimiter.
13.  Load the `star_distances.csv` file back into a NumPy array and print it.

## Summary

Congratulations! You've taken your first steps into using NumPy, a powerful library for numerical operations in Python. You've learned about:

*   The importance of NumPy for efficient numerical calculations, especially with arrays.
*   The core `ndarray` object.
*   Creating 1D and 2D arrays.
*   Checking array attributes (`.shape`, `.size`, `.ndim`, `.dtype`).
*   Accessing and modifying array elements using indexing and slicing.
*   Performing element-wise arithmetic and comparison operations on arrays.
*   Using common NumPy functions like `sum()`, `mean()`, `max()`, and `min()`.
*   Saving and loading arrays using `np.save()`, `np.load()`, `np.savetxt()`, `np.loadtxt()` and `np.genfromtxt()`.

NumPy is a foundational tool in the scientific Python ecosystem. As you continue your journey in astronomy and programming, you'll find it indispensable for analysing data, running simulations, visualising results and saving and loading arrays. Keep practising these fundamental concepts!

**Additional Resources**

Numpy user guide: <https://numpy.org/doc/stable/user/index.html>

Numpy for beginners: <https://numpy.org/doc/stable/user/absolute_beginners.html>

Numpy arrays: <https://www.datacamp.com/tutorial/python-numpy-tutorial>

Vectors and matrices: <https://www.kaggle.com/code/jhossain/introduction-to-numpy-vectors-and-matrices>